spark/0001-Add-cook-support-for-spark-v1.6.1.patch

```diff
@@ -0,0 +1,769 @@
++                       .setExtract(uri.getExtract)
++                       .setExecutable(uri.getExecutable)
++                       .build()
++       }
++    logDebug(s"command: $commandString")
++
++    val remoteHdfsConf = conf.get("spark.executor.cook.hdfs.conf.remote", "")
++    val remoteConfFetch = if (remoteHdfsConf.nonEmpty) {
++      val name = Paths.get(remoteHdfsConf).getFileName
++      Seq(
++        fetchUri(remoteHdfsConf),
++        "mkdir HADOOP_CONF_DIR",
++        s"tar --strip-components=1 -xvzf $name -C HADOOP_CONF_DIR",
++        // This must be absolute because we cd into the spark directory
++        s"export HADOOP_CONF_DIR=`pwd`/HADOOP_CONF_DIR",
++        "export HADOOP_CLASSPATH=$HADOOP_CONF_DIR"
++      )
++
++    } else Seq()
++
++    val cleanup = "if [ -z $KEEP_SPARK_LOCAL_DIRS ]; then rm -rf $SPARK_LOCAL_DIRS; echo deleted $SPARK_LOCAL_DIRS; fi"
++
++    val cmds = remoteConfFetch ++ environment ++ Seq(commandString, cleanup)
++
++    new Job.Builder()
++      .setUUID(jobId)
++      .setEnv(environment.asJava)
++      .setUris(uris.asJava)
++      .setCommand(cmds.mkString("; "))
++      .setMemory(mesosBackend.calculateTotalMemory(sc))
++      .setCpus(numCores)
++      .setPriority(priority)
++      .build()
++  }
++
++  def createRemainingJobs(): List[Job] = {
++    @annotation.tailrec
++    def loop(coresRemaining: Int, jobs: List[Job]): List[Job] =
++      if (coresRemaining <= 0) {
++        jobs
++      } else if (coresRemaining <= maxCoresPerJob) {
++        createJob(coresRemaining) :: jobs
++      } else {
++        loop(coresRemaining - maxCoresPerJob, createJob(maxCoresPerJob) :: jobs)
++      }
++    loop(currentCoresLimit, List()).reverse
++  }
++
++  def requestRemainingCores() : Unit = {
++    val jobs = createRemainingJobs()
++
++    totalCoresRequested += jobs.map(_.getCpus.toInt).sum
++    runningJobUUIDs = runningJobUUIDs ++ jobs.map(_.getUUID)
++
++    if (!jobs.isEmpty) { jobClient.submit(jobs.asJava, jobListener) }
++  }
++
++  def currentCoresLimit(): Int =
++    jobLimit.map(_ * maxCoresPerJob).getOrElse(maxCores) - totalCoresRequested
++
++  override def doKillExecutors(executorIds: Seq[String]): Boolean =
++    abortJobs(executorIds.map(x => UUID.fromString(x)).toList)
++
++  override def doRequestTotalExecutors(requestedTotal: Int): Boolean = {
++    logInfo(s"Capping the total amount of executors to $requestedTotal")
++    jobLimit = Some(requestedTotal)
++    requestRemainingCores
++    true
++  }
++
++  // todo we should consider making this async, because if there are issues with cook it'll
++  // just block the spark shell completely. We can block if they submit something (we don't
++  // want to get into thread management issues)
++  override def start() {
++    super.start()
++    logInfo("Starting Cook Spark Scheduler")
++    requestRemainingCores()
++  }
++
++  override def stop() {
++    super.stop()
++    abortJobs(runningJobUUIDs.toList)
```

Comment:
We *probably* should not be aborting jobs here, as it results in killing all jobs that the cook scheduler thinks are still running. I've tested it w/o this line and the behavior seems to be correct.